### PR TITLE
maintenance: statusMember() and defragmentMember() now takes a endpoint

### PR DESCRIPTION
**Pull request by fanminshi**

### CODE REVIEW

**Comment by fanminshi** [comment]
Work in Progress.

File: src/main/java/com/coreos/jetcd/MaintenanceImpl.java [hunk 1]
```diff
@@ -38,10 +43,34 @@
...
+        .configureStub(MaintenanceGrpc.newFutureStub(c.getChannel()), c.getToken());
+    this.streamStub = ClientUtil
+        .configureStub(MaintenanceGrpc.newStub(c.getChannel()), c.getToken());
+  }
+
+  @FunctionalInterface
+  private interface DialFunction {
+
+    /**
+     * dial dials to an endpoint and returns a managed channel its associated token.
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/MaintenanceImpl.java]: a managed channel and its associated token

File: src/main/java/com/coreos/jetcd/MaintenanceImpl.java [hunk 2]
```diff
@@ -38,10 +43,34 @@
...
+     *
+     * @return a managed channel its associated token.
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/MaintenanceImpl.java]: and its associated token

Comment (xiang90) [inline src/main/java/com/coreos/jetcd/MaintenanceImpl.java]: i am not familiar with gRPC java. so a stupid question, why the token is needed?

File: src/main/java/com/coreos/jetcd/Client.java [hunk 3]
```diff
@@ -45,30 +63,41 @@ public Client(ManagedChannelBuilder<?> channelBuilder, ClientBuilder clientBuild
...
 
+  Pair<ManagedChannel, Optional<String>> toChannelAndToken(ManagedChannelBuilder<?> channelBuilder)
+      throws AuthFailedException, ConnectException {
+    checkNotNull(channelBuilder, "channelBuilder can't be null");
+    ManagedChannel managedChannel = channelBuilder.build();
+    Optional<String> token = this.generateToken(managedChannel, this.user, this.pass);
+    return new Pair<>(managedChannel, token);
+  }
+
+  // ************************
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/Client.java]: remove this line.

File: src/main/java/com/coreos/jetcd/MaintenanceImpl.java (hunk 2 above)
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/MaintenanceImpl.java]: is this the authToken?

Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/MaintenanceImpl.java]: the token here is same as the one in go client counter part at https://github.com/coreos/etcd/blob/master/clientv3/client.go#L286.  If auth is enabled for etcd, the this token is needed to build a stub that has the token as its metadata. then this token is passed to etcd server for authentication.

File: src/main/java/com/coreos/jetcd/data/Pair.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
+package com.coreos.jetcd.data;
```
Comment (xiang90) [inline src/main/java/com/coreos/jetcd/data/Pair.java]: not sure if this belongs to here. this is not actually etcd kv pair but some internal helper structure, right? i feel it is confusing to put it under data.

**Comment by xiang90** [comment]
LGTM

File: src/main/java/com/coreos/jetcd/data/Pair.java (hunk 4 above)
Comment (fanminshi) [inline src/main/java/com/coreos/jetcd/data/Pair.java]: agreed. pair is merely a helper structure. I probaly have another pkg called `com.coreos.jetcd.internal` and put pair there?

Comment (xiang90) [inline src/main/java/com/coreos/jetcd/data/Pair.java]: sgtm

**Pull request merged**